### PR TITLE
feat(dataservices): metadata quality score

### PR DESCRIPTION
**Pull request by Samuelfaure**
## Summary

Adds a metadata quality score to dataservices, mirroring the existing dataset quality score.

- New `quality_cached` field + `compute_quality()` computing 10 boolean criteria: `has_description`, `has_machine_documentation`, `has_technical_documentation`, `has_business_documentation`, `license`, `has_contact_point`, `has_base_api_url`, `availability_documented`, `rate_limiting_documented`, `access_conditions_clear`.
- `normalize_score()` / `compute_quality_score()` produce a `0–1` score (count of satisfied criteria ÷ 10).
- A `quality` property (criteria + `score`) is exposed on the dataservice API via `dataservice_quality_fields`.
- `clean()` recomputes the cache on every save (same pattern as `Dataset.clean()`).
- Migration `2026-06-05-dataservices-quality-cached.py` backfills existing dataservices (optional — the `quality` property recomputes on read when the cache is empty).

`access_conditions_clear` is satisfied automatically when `access_type` is `OPEN`; otherwise it requires `authorization_request_url`.

## Test Plan

- [x] `pytest udata/tests/dataservice/test_dataservice_quality.py` — 12 tests (criteria, scoring, property, API serialization, save-hook cache, edge cases for `availability=0.0` and `OPEN_WITH_ACCOUNT`)
- [x] `pytest udata/tests/api/test_dataservices_api.py udata/tests/apiv2/test_dataservices.py` — no regressions (36 passed total)
- [ ] Run the backfill migration on deploy: `udata db migrate`

closes https://linear.app/pole-api/issue/API2-354/mettre-en-place-des-indicateurs-de-qualite-sur-les-fiches-api

### CODE REVIEW

**Comment by maudetes** [comment]
Thank you ! Indeed, a metadata quality seems like a good incentive for publisher to improve their metadata description!

@agarrone if you want to take a look at this, in particular the criteria listed here

>    computing 10 boolean criteria: `has_description`, `has_machine_documentation`, `has_technical_documentation`, `has_business_documentation`, `license`, `has_contact_point`, `has_base_api_url`, `availability_documented`, `rate_limiting_documented`, `access_conditions_clear`.

**Comment by ThibaudDauce** [comment]
What do @agarrone think of this? :-)

**Comment by Samuelfaure** [comment]
Still missing: need to revisit criteria weighting with @DorineLam

**Comment by ThibaudDauce** [comment]
@agarrone we're waiting for your opinion on the subject

**Comment by agarrone** [comment]
Sorry for the late reply and thanks for this contribution. Since we plan to reshape how the score works for datasets in the following month, I prefer not to implement a score for dataservices that might change soon.
But we will definitely use this to discuss how it could work on the dataservices

@maudetes do you have thoughts on this?

**Comment by Samuelfaure** [comment]
@agarrone The calculation on dataservice / dataset is completely different so I'd wager we could just ship this 

But I understand if you prefer to redo dataset before, will it be done in 2026?

**Comment by maudetes** [comment]
@agarrone, in my opinion, it would indeed be better to update datasets score before deriving it to dataservices (in order to share the same target vocabulary).
But it depends on the roadmap as @Samuelfaure says and I guess the that having a score quality showed to producers would nudge them to improve on their description, and sooner is better?

**Comment by Samuelfaure** [comment]
Note: [e2e (chromium, 2) CI failure unrelated to this PR, fixed in https://github.com/datagouv/cdata/pull/1177#event-28723760001

**Comment by JeSuisUnCaillou** [comment]
Des news de l'avancée de ce sujet ? @agarrone

File: udata/core/dataservices/models.py [hunk 1]
```diff
@@ -406,6 +438,63 @@ def count_reuses(self):
...
+                "has_description": 2,
+                "rate_limiting_documented": 2,
+                "has_base_api_url": 2,
+                "has_technical_documentation": 2,
+                "access_conditions_clear": 1,
+                "license": 1,
+                "availability_documented": 1,
+                "has_contact_point": 1,
+                "has_business_documentation": 1,
+            }
```
Comment (ThibaudDauce) [inline udata/core/dataservices/models.py:493]: A dataservice without swagger can score 100% here since the key leaves `weights`
entirely, while the tooltip shows it as missing right below the bar. Keep the weight in
the total?

File: udata/core/dataservices/models.py [hunk 2]
```diff
@@ -406,6 +438,63 @@ def count_reuses(self):
...
+        result = {}
+
+        result["has_description"] = len(self.description or "") > current_app.config.get(
+            "QUALITY_DESCRIPTION_LENGTH"
+        )
+        result["has_machine_documentation"] = bool(self.machine_documentation_url)
+        result["has_technical_documentation"] = bool(self.technical_documentation_url)
+        result["has_business_documentation"] = bool(self.business_documentation_url)
+        result["license"] = bool(self.license)
+        result["has_contact_point"] = bool(self.contact_points)
```
Comment (ThibaudDauce) [inline udata/core/dataservices/models.py:452]: `contact_points` has `reverse_delete_rule=PULL`, and mongoengine applies it with a
queryset `update(pull_all__…)`, so no `save()` and no `clean()`. Delete the last contact
point and `quality_cached["has_contact_point"]` stays `true` until someone edits the
dataservice again. Recompute it on the cascade?

File: udata/core/dataservices/models.py [hunk 3]
```diff
@@ -406,6 +438,63 @@ def count_reuses(self):
...
+                "access_conditions_clear": 1,
+                "license": 1,
+                "availability_documented": 1,
+                "has_contact_point": 1,
+                "has_business_documentation": 1,
+            }
+        else:
+            # has_machine_documentation absent: when swagger is missing it contributes nothing,
+            # so base_api_url and technical_documentation take higher weight as fallbacks
+            weights = {
```
Comment (ThibaudDauce) [inline udata/core/dataservices/models.py:483]: Both tables repeat nine identical entries and only three change
(`has_machine_documentation`, `has_base_api_url`, `has_technical_documentation`). Could
the second one derive from the first so the delta is readable?

File: udata/core/dataservices/models.py [hunk 4]
```diff
@@ -406,6 +438,63 @@ def count_reuses(self):
...
+                "has_base_api_url": 2,
+                "has_technical_documentation": 2,
+                "access_conditions_clear": 1,
+                "license": 1,
+                "availability_documented": 1,
+                "has_contact_point": 1,
+                "has_business_documentation": 1,
+            }
+        total = sum(weights.values())
+        score = sum(weights[k] for k, v in quality.items() if v is True and k in weights)
```
Comment (ThibaudDauce) [inline udata/core/dataservices/models.py:495]: Iterating on `weights` makes both guards useless, no?
```suggestion
        score = sum(weight for key, weight in weights.items() if quality.get(key))
```

File: udata/migrations/2026-06-05-dataservices-quality-cached.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+def migrate(db):
+    log.info("Backfilling Dataservice.quality_cached...")
+
+    with click.progressbar(
+        Dataservice.objects(), length=Dataservice.objects().count()
+    ) as dataservices:
+        for dataservice in dataservices:
+            try:
+                dataservice.quality_cached = dataservice.compute_quality()
+                dataservice.save(signal_kwargs={"ignores": ["post_save"]})
```
Comment (ThibaudDauce) [inline udata/migrations/2026-06-05-dataservices-quality-cached.py:23]: `save()` runs `clean()`, which already sets `quality_cached`. The dataset migration just
calls `dataset.save()` for that reason.
```suggestion
                dataservice.save(signal_kwargs={"ignores": ["post_save"]})
```

**Review by ThibaudDauce — CHANGES_REQUESTED** [review]
The description still says `normalize_score()` and a `count of satisfied criteria ÷ 10`,
but the code weights the criteria on totals of 14 and 13. Could you update it?

---

Could you also add the quality to the dataservice card mask?

```suggestion
DATASERVICE_CARD_MASK = "{id,title,acronym,slug,description,base_api_url,format,self_api_url,self_web_url,organization,owner,quality,metrics,badges,tags,private,created_at,metadata_modified_at}"
```
Otherwise the quality bar shows up on dataset cards and silently disappears on
dataservice ones in the same bloc, since `DataserviceCard` guards it with a `v-if`.

---

Datasets export their score here (`dataset/csv.py:61`). Add it for dataservices too?
```suggestion
        ("harvest.last_update", lambda r: r.harvest and r.harvest.last_update),
        ("quality_score", lambda o: format(o.quality["score"], ".2f")),
    )
```
